lib/spack/docs/spec_syntax.rst

```diff
@@ -0,0 +1,712 @@
+  ``--qt``, or ``~~qt``).
+* ``name=<value>`` Optional variant specifiers that are not restricted to
+  boolean variants. Use ``name==<value>`` to propagate variant through the
+  dependencies.
+* ``name=<value>`` Optional compiler flag specifiers. Valid flag names are
+  ``cflags``, ``cxxflags``, ``fflags``, ``cppflags``, ``ldflags``, and ``ldlibs``.
+  Use ``name==<value>`` to propagate compiler flags through the dependencies.
+* ``target=<value> os=<value>`` Optional architecture specifier
+  (e.g., ``target=haswell os=CNL10``)
+* ``%`` Direct dependency specs. Specs the user knows are not merely present in
+  the graph, but depended on directly by the previous node.
+* ``^`` Dependency specs (e.g., ``^callpath@1.1``). These dependencies may appear
+  anywhere in the link/run dependencies of the root, or in the direct build
+  dependencies.
+
+There are two things to notice here. The first is that specs are
+recursively defined. That is, each dependency after ``%`` or ``^`` is
+a spec itself. The second is that everything is optional *except* for
+the initial package name identifier. Users can be as vague or as
+specific as they want about the details of building packages, and this
+makes Spack good for beginners and experts alike.
+
+To really understand what's going on above, we need to think about how
+software is structured. An executable or a library (these are
+generally the artifacts produced by building software) depends on
+other libraries in order to run. We can represent the relationship
+between a package and its dependencies as a graph. Here is the full
+dependency graph for ``mpileaks``:
+
+.. graphviz::
+
+   digraph {
+       mpileaks -> mpich
+       mpileaks -> callpath -> mpich
+       callpath -> dyninst
+       dyninst  -> libdwarf -> libelf
+       dyninst  -> libelf
+   }
+
+Each box above is a package, and each arrow represents a dependency on
+some other package. For example, we say that the package ``mpileaks``
+*depends on* ``callpath`` and ``mpich``. ``mpileaks`` also depends
+*indirectly* on ``dyninst``, ``libdwarf``, and ``libelf``, in that
+these libraries are dependencies of ``callpath``. To install
+``mpileaks``, Spack has to build all of these packages. Dependency
+graphs in Spack have to be acyclic, and the *depends on* relationship
+is directional, so this is a *directed, acyclic graph* or *DAG*.
+
+The package name identifier in the spec is the root of some dependency
+DAG, and the DAG itself is implicit. Spack knows the precise
+dependencies among packages, but users do not need to know the full
+DAG structure. Each ``^`` in the full spec refers to some dependency
+of the root package. Spack will raise an error if you supply a name
+after ``^`` that the root does not actually depend on (e.g., ``mpileaks
+^emacs@23.3``). Each ``%`` refers to some direct dependency, and Spack will
+similarly raise an error if that relationship is invalid.
```

Comment:
why so much focus on "raising an error" :thinking: it's about the meaning of ^ and %?